hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java

```diff
@@ -0,0 +1,263 @@
+
+    private final NodeEngineImpl nodeEngine;
+    private final ILogger logger;
+    private final MetricsConfig config;
+    private final LiveOperationRegistry liveOperationRegistry;
+    // Holds futures for pending read metrics operations
+    private final ConcurrentMap<CompletableFuture<RingbufferSlice<Map.Entry<Long, byte[]>>>, Long>
+            pendingReads = new ConcurrentHashMap<>();
+    private final ProbeRenderer probeRenderer = new PublisherProbeRenderer();
+
+    /**
+     * Ringbuffer which stores a bounded history of metrics. For each round of collection,
+     * the metrics are compressed into a blob and stored along with the timestamp,
+     * with the format (timestamp, byte[])
+     */
+    private ConcurrentArrayRingbuffer<Map.Entry<Long, byte[]>> metricsJournal;
+    private volatile ScheduledFuture<?> scheduledFuture;
+
+    private final List<MetricsPublisher> publishers;
+
+    private final Supplier<MetricsRegistry> metricsRegistrySupplier;
+
+    public MetricsService(NodeEngine nodeEngine) {
+        this(nodeEngine, ((NodeEngineImpl) nodeEngine)::getMetricsRegistry);
+    }
+
+    public MetricsService(NodeEngine nodeEngine, Supplier<MetricsRegistry> metricsRegistrySupplier) {
+        this.nodeEngine = (NodeEngineImpl) nodeEngine;
+        this.logger = nodeEngine.getLogger(getClass());
+        this.config = nodeEngine.getConfig().getMetricsConfig();
+        this.liveOperationRegistry = new LiveOperationRegistry();
+        this.metricsRegistrySupplier = metricsRegistrySupplier;
+        this.publishers = new CopyOnWriteArrayList<>();
+    }
+
+    @Override
+    public void init(NodeEngine nodeEngine, Properties properties) {
+        this.publishers.addAll(getPublishers());
+
+        if (publishers.isEmpty()) {
+            return;
+        }
+
+        logger.fine("Configuring metrics collection, collection interval=" + config.getCollectionIntervalSeconds()
+                + " seconds, retention=" + config.getRetentionSeconds() + " seconds, publishers="
+                + publishers.stream().map(MetricsPublisher::name).collect(joining(", ", "[", "]")));
+
+        ExecutionService executionService = nodeEngine.getExecutionService();
+        scheduledFuture = executionService.scheduleWithRepetition("MetricsPublisher", this::collectMetrics, 1,
+                config.getCollectionIntervalSeconds(), TimeUnit.SECONDS);
+    }
+
+    /**
+     * Register a custom {@link MetricsPublisher} implementation with a register
+     * function that takes {@link NodeEngine} as input letting the caller to
+     * optionally initialize the publisher returned from the function.
+     *
+     * @param registerFunction The function that returns with the {@link MetricsPublisher}
+     *                         instance.
+     */
+    public void registerPublisher(Function<NodeEngine, MetricsPublisher> registerFunction) {
+        MetricsPublisher publisher = registerFunction.apply(nodeEngine);
```

Comment:
this function should only be called if metrics are enabled.

Reply:
Done. We log on `fine` level if the metrics system is disabled.